Fix issue #14457: Contact names with ampersands not showing in search

goma101 opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 7.x)
* a.b for any bug fixes (e.g. 5.2, 6.0)
* c.x for any bug fixes, features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 7.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ✔️ <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | ❌
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️ <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | N/A <!-- required for new features -->
| Related developer documentation PR URL | N/A <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #14457 <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

This PR fixes an issue where a global search would not return contacts (or other entities) with ampersands in one of their fields such as their names, if searched with the ampersand. This bug occurred because the `InputHelper::clean` method encodes characters such as ampersand into html encoding, thus causing a mismatch when executing a global search.

This fix decodes search strings before global search execution, thereby eliminating this mismatch and yielding correct results. The PR also includes a phpunit test, for the search example provided in the issue (Global search for Contact with name "R&D"), asserting the search is successful.


<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->

| Before                                 | After
| -------------------------------------- | ---
| ![before](https://github.com/user-attachments/assets/d308b587-de79-415e-85d8-87a92de1c7c1) | ![after](https://github.com/user-attachments/assets/aa59ae9b-e5a1-41ff-b9a8-6ae6c6737758)



---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a contact or other entity with "&" in one of its fields
3. Execute a global search for the contact/entity by that field, **including the &** in the search
4. Confirm that the results are correct.

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

goma101 committed:
fix #14457: decode special characters in searches

fixes searches for fields with characters otherwise encoded like
ampersand

adds test to check ampersand shows in search, on created contact

patrykgruszka reviewed (changes requested):
I suggest removing the `InputHelper::clean` in the controller instead of decoding the query in every subscriber.

goma101 committed:
fix #14457: remove InputHelper::clean from global search action

reverts previous individual search subscriber modification, removes
clean method from global search action in ajax controller

goma101 commented:
@patrykgruszka Indeed much more sensical, my bad.

Please tell me if it needs anything else, or if you'd like me to squash or something.

patrykgruszka commented on app/bundles/LeadBundle/Tests/Functional/SearchTestHelper.php:
```diff
@@ -0,0 +1,44 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+abstract class SearchTestHelper extends MauticMysqlTestCase
```
```suggestion
abstract class AbstractSearchTest extends MauticMysqlTestCase
```

To maintain consistency, we should adjust the class name. Helper classes should be used as dependencies rather than abstract classes.

There are two options:
Option 1: Rename the class to something like `AbstractSearchTest` and it should be fine
Option 2: Modify it to function as a helper, similar to `app/bundles/EmailBundle/Tests/Functional/Fixtures/EmailFixturesHelper.php`.

goma101 replied:
I'll rename, and if you agree, I think it would also be better for extensibility to move the remaining search-related auxiliary functions in `SearchWithCustomFieldDataFunctionalTest.php` to the abstract.

patrykgruszka replied:
Looks good!

goma101 committed:
refactor SearchTestHelper into AbstractSearchTest

renames SearchTestHelper and moves auxiliary functions from the subclass
to the parent

patrykgruszka reviewed (approved):
The code looks fine and fixes the problem :+1: 
![image](https://github.com/user-attachments/assets/2935d7c4-b588-4ea7-b4f1-b3d4cdd2bc8a)

escopecz reviewed (approved):
I also believe that removing `InputHelper::clean` in this case is the right way to solve this 👍
